src/sentry/snuba/metrics/metric_extraction.py

```diff
@@ -0,0 +1,258 @@
+        if not cls.check(field, query):
+            return None
+
+        relay_field, metric_type, op = _extract_field_info(field)
+
+        return cls(
+            _field=field,
+            _query=query,
+            field=relay_field,
+            metric_type=metric_type,
+            mri=f"{metric_type}:{CUSTOM_ALERT_METRIC_NAME}@none",
+            op=op,
+        )
+
+    def query_hash(self) -> str:
+        str_to_hash = f"{self.field};{self._query}"
+        return hashlib.shake_128(bytes(str_to_hash, encoding="ascii")).hexdigest(4)
+
+    def condition(self) -> RuleCondition:
+        where, having = _get_query_builder().resolve_conditions(self._query, False)
+
+        assert where, "Query should not use on demand metrics"
+        assert not having, "Having conditions are not supported"
+
+        where = [c for c in (_convert_condition(c) for c in where) if c is not None]
+
+        if len(where) == 1:
+            return where[0]
+        else:
+            return {"op": "and", "inner": where}
+
+
+def _extract_field_info(aggregate: str) -> Tuple[Optional[str], str, MetricOperationType]:
+    select = _get_query_builder().resolve_column(aggregate, False)
+
+    metric_type = _AGGREGATE_TO_METRIC_TYPE.get(select.function)
+    metric_op = _SNUBA_TO_METRIC_AGGREGATES.get(select.function)
+    assert metric_type and metric_op, f"Unsupported aggregate function {select.function}"
+
+    if metric_type == "c":
+        assert not select.parameters, "Count should not have parameters"
+        return None, metric_type, metric_op
+    else:
+        assert len(select.parameters) == 1, "Only one parameter is supported"
+
+        name = select.parameters[0].name
+        assert name in _SNUBA_TO_RELAY_FIELDS, f"Unsupported field {name}"
+
+        return _SNUBA_TO_RELAY_FIELDS[name], metric_type, metric_op
+
+
+def _convert_condition(condition: Union[Condition, BooleanCondition]) -> Optional[RuleCondition]:
+    if isinstance(condition, BooleanCondition):
+        return cast(
+            RuleCondition,
+            {
+                "op": condition.op.name.lower(),
+                "inner": [_convert_condition(c) for c in condition.conditions],
+            },
+        )
+
+    assert isinstance(condition, Condition), f"Unsupported condition type {type(condition)}"
+
+    # TODO: Currently we do not support function conditions like count_if
+    if not isinstance(condition.lhs, Column):
+        return None
+
+    assert condition.lhs.name in _SNUBA_TO_RELAY_FIELDS, f"Unsupported field {condition.lhs.name}"
+
+    return {
+        "op": condition.op.name.lower(),
+        "name": _SNUBA_TO_RELAY_FIELDS[condition.lhs.name],
+        "value": condition.rhs,
+    }
+
+
+def _get_query_builder():
+    from sentry.search.events.builder import QueryBuilder
+
+    return QueryBuilder(
+        dataset=Dataset.Transactions,
+        params={"start": datetime.now(), "end": datetime.now()},
```

Comment:
Are these just added for the sake of not making the instantiation fail?

Reply:
Correct. QueryBuilder requires some time range which we then ignore. Even if we pass a `query` and then access QueryBuilder's `where`, the time range is not used.

As your comment above mentions, we will have to revisit how to parse queries and fields, so view this as a proof of concept for vertical integration.

Reply:
For now, I added a comment just so that it is clear that the parameters are not actually used.